src/backend/distributed/commands/create_citus_local_table.c

```diff
@@ -0,0 +1,620 @@
+ * single placement is only allowed to be on the coordinator.
+ */
+Datum
+create_citus_local_table(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	Oid relationId = PG_GETARG_OID(0);
+
+	/*
+	 * Lock target relation with an exclusive lock - there's no way to make
+	 * sense of this table until we've committed, and we don't want multiple
+	 * backends manipulating this relation.
+	 */
+	Relation relation = try_relation_open(relationId, ExclusiveLock);
+
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("could not create citus local table: "
+							   "relation does not exist")));
+	}
+
+	ErrorIfUnsupportedCreateCitusLocalTable(relationId);
+
+	ObjectAddress tableAddress = { 0 };
+	ObjectAddressSet(tableAddress, RelationRelationId, relationId);
+
+	/*
+	 * Ensure dependencies first as we will create shell table on the other nodes
+	 * in the MX case.
+	 */
+	EnsureDependenciesExistOnAllNodes(&tableAddress);
+
+	CreateCitusLocalTable(relationId);
+
+	relation_close(relation, NoLock);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * ErrorIfUnsupportedCreateCitusLocalTable errors out if we cannot create the
+ * citus local table from the relation relationId.
+ */
+static void
+ErrorIfUnsupportedCreateCitusLocalTable(Oid relationId)
+{
+	/* citus local tables can only be created from coordinator for now */
+	EnsureCoordinator();
+
+	EnsureTableOwner(relationId);
+
+	/* we allow creating citus local tables only from relations with RELKIND_RELATION */
+	EnsureRelationKindSupported(relationId);
+
+	if (PartitionTable(relationId) || PartitionedTable(relationId))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("citus local tables can not be involved in a "
+							   "partition relationship")));
+	}
+
+	EnsureTableNotDistributed(relationId);
+
+	if (!CoordinatorAddedAsWorkerNode())
+	{
+		const char *relationName = get_rel_name(relationId);
+
+		Assert(relationName != NULL);
+
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg(
+							"cannot create citus local table \"%s\", citus local "
+							"tables can only be created from coordinator node if "
+							"it is added to pg_dist_node", relationName)));
+	}
+}
+
+
+/*
+ * CreateCitusLocalTable is the internal method to create citus local table's
+ * shard and metadata.
+ * Note that this function does not perform any validation on the table as it
+ * is already done in create_citus_local_table.
+ */
+static void
+CreateCitusLocalTable(Oid shellRelationId)
+{
+	/*
+	 * Make sure that existing reference tables have been replicated to all
+	 * the nodes such that we can create foreign keys and joins work
+	 * immediately after creation.
+	 */
+	EnsureReferenceTablesExistOnAllNodes();
+
+	/*
+	 * Get an exclusive lock on relation with shellRelationId as the operations
+	 * done in this function should be automic.
+	 */
+	LockRelationOid(shellRelationId, ExclusiveLock);
```

Comment:
The reason we take lock there is exactly what you wrote:

> because in the future this method can be called from different places other than create_citus_local_table?

As it may be called from somewhere else in future, we should ensure that table is locked against parallel modifications even if we are using the internal method.

Reply:
So in the future if a caller gets a lock different than `ExclusiveLock` it will go into a deadlock right?

Reply:
Afaik, it shouldn't be the case as the caller and the `CreateCitusLocalTable` executes in the same transaction. (But still I should validate my knowledge)

Reply:
(Now that I pushed all the logic down into the internal method, we do not acquire any lock in udf)